functions/diff.js

```diff
@@ -1,39 +1,53 @@
 const parse = require('./parse')
 const eq = require('./eq')
+const gt = require('./gt')
 
 const diff = (version1, version2) => {
   const v1 = parse(version1)
   const v2 = parse(version2)
   if (eq(v1, v2)) {
     return null
-  } else {
-    const hasPre = v1.prerelease.length || v2.prerelease.length
-    const prefix = hasPre ? 'pre' : ''
-    const defaultResult = hasPre ? 'prerelease' : ''
-
-    if (v1.major !== v2.major) {
-      return prefix + 'major'
-    }
-    if (v1.minor !== v2.minor) {
-      return prefix + 'minor'
-    }
-
-    if (v1.patch !== v2.patch) {
-      return prefix + 'patch'
-    }
-
-    if (!v1.prerelease.length || !v2.prerelease.length) {
-      if (v1.patch) {
-        return 'patch'
-      }
-      if (v1.minor) {
-        return 'minor'
-      }
-      if (v1.major) {
-        return 'major'
-      }
-    }
-    return defaultResult // may be undefined
   }
+
+  const v1Higher = gt(v1, v2)
+  const highVersion = v1Higher ? v1 : v2
+  const lowVersion = v1Higher ? v2 : v1
+  const highHasPre = !!highVersion.prerelease.length
+  const lowHasPre = !!lowVersion.prerelease.length
+  const prefix = highHasPre && !lowHasPre ? 'pre' : ''
```

Comment:
This is where a lot of magic now lies and could definitely use a comment.

Reply:
I’ve added one. Look ok?

Reply:
See that's exactly why we need that comment.  In my mind "high" prioritizes before "low" but semver is considering going from low to high.  Great comment, thanks.